### PR TITLE
Add robust cost function

### PR DESCRIPTION
**Pull request by fantaosha**
## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: theseus/core/robust_cost_function.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        return self.cost_function.error()
+
+    def jacobians(self) -> Tuple[List[torch.Tensor], torch.Tensor]:
+        return self.cost_function.jacobians()
+
+    def weighted_error(self) -> torch.Tensor:
+        return self.cost_function.weighted_error()
+
+    def weighted_jacobians_error(self) -> Tuple[List[torch.Tensor], torch.Tensor]:
+        return self.weighted_jacobians_error()
```
Comment (mhmukadam) [inline theseus/core/robust_cost_function.py:39]: Is this returning itself? Could be a bug.

File: theseus/core/robust_cost_function.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+from .loss import Loss
+
+
+class RobustCostFunction(th.CostFunction):
+    def __init__(
+        self,
+        loss_function: Loss,
+        log_loss_radius: th.Vector,
+        cost_function: th.CostFunction,
```
Comment (mhmukadam) [inline theseus/core/robust_cost_function.py:20]: Is the idea of this implementation that any cost function can be turned into its robust version?

File: theseus/core/robust_cost_function.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+from typing import List, Tuple
+
+import torch
+
+import theseus as th
+
+from .loss import Loss
+
+
+class RobustCostFunction(th.CostFunction):
```
Comment (mhmukadam) [inline theseus/core/robust_cost_function.py:15]: Add unit testing for this.

File: theseus/core/robust_cost_function.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    def weighted_jacobians_error(self) -> Tuple[List[torch.Tensor], torch.Tensor]:
+        return self.weighted_jacobians_error()
+
+    def value(self) -> torch.Tensor:
+        weighted_error = self.weighted_error()
+        squared_norm = torch.sum(weighted_error**2, dim=1, keepdim=True)
+        loss_radius = torch.exp(self.log_loss_radius.data)
+        return self.loss_function.evaluate(squared_norm, loss_radius)
+
+    def rescaled_jacobians_error(self) -> Tuple[List[torch.Tensor], torch.Tensor]:
```
Comment (mhmukadam) [inline theseus/core/robust_cost_function.py:47]: I am wondering if we can simply the API here and avoid introducing this new function.

If `weighted_jacobians_error()` is overridden for `RobustCostFunction` to do what `rescaled_jacobians_error()` is doing then you don't need a separate new function for this.

File: theseus/core/robust_cost_function.py (hunk 2 above)
Comment (fantaosha) [reply theseus/core/robust_cost_function.py:20]: Yes.

File: theseus/core/robust_cost_function.py (hunk 3 above)
Comment (fantaosha) [reply theseus/core/robust_cost_function.py:15]: Will do. I'm not considering to merge this branch to main. This is only for profiling and benchmarking.

File: theseus/core/robust_cost_function.py (hunk 4 above)
Comment (fantaosha) [reply theseus/core/robust_cost_function.py:47]: Since this feature is not ready to merge to main and has not been fully tested, I just implemented another function for simplification.

File: theseus/core/objective.py [hunk 5]
```diff
@@ -348,15 +348,13 @@ def error(
                 for var in self.optim_vars:
                     old_data[var] = self.optim_vars[var].data
             self.update(input_data=input_data)
-        error_vector = torch.zeros(self.batch_size, self.dim()).to(
+        error_vector = torch.zeros(self.batch_size, len(self.cost_functions)).to(
```
Comment (mhmukadam) [inline theseus/core/objective.py:351]: Does `len(self.cost_functions)` give the number (n) of cost functions? I think we want \sum_i{cost_i.dim}, where i=1,..,n.

File: theseus/core/loss.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+import abc
+
+import torch
+
+
+class Loss:
```
Comment (mhmukadam) [inline theseus/core/loss.py:11]: Loss might be a confusing name here given the outer-loop training and learning related terminology.

Also if this is only relevant to `RobustCostFunction` we can probably move all this to that file.

**Review by mhmukadam — COMMENTED** [review]
Left some comments and discussion points.

File: theseus/core/loss.py (hunk 6 above)
Comment (fantaosha) [reply theseus/core/loss.py:11]: How about `LossFunction`?

File: theseus/core/cost_function.py [hunk 7]
```diff
@@ -57,6 +57,13 @@ def weighted_jacobians_error(
         jacobian, err = self.jacobians()
         return self.weight.weight_jacobians_and_error(jacobian, err)
 
+    def value(self) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (fantaosha) [reply theseus/core/cost_function.py:60]: It calculates the subobjective value of `CostFunction`.

File: theseus/core/loss.py (hunk 6 above)
Comment (mhmukadam) [reply theseus/core/loss.py:11]: `RobustCost`?
That way `RobustCost` + `CostFunction` -> `RobustCostFunction` would make sense too.

**Comment by fantaosha** [comment]
@luisenp @mhmukadam I completely rewrote robust cost functions. I merged `RobustCostFunction` to `CostFunction`. Maybe we need a discussion about this.

**Comment by fantaosha** [comment]
@luisenp @mhmukadam I completely rewrote robust cost functions. I merged `RobustCostFunction` to `CostFunction`. Maybe we need a discussion about this.

File: theseus/core/loss_function.py [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+        pass
+
+    @abc.abstractmethod
+    def rescale(
+        self,
+        jacobians: List[torch.Tensor],
+        error: torch.Tensor,
+    ) -> Tuple[List[torch.Tensor], torch.Tensor]:
+        pass
+
```
Comment (mhmukadam) [inline theseus/core/loss_function.py:31]: Add a `_copy_impl` abstract method here like [this](https://github.com/facebookresearch/theseus/blob/main/theseus/core/cost_function.py#L62).

File: theseus/core/loss_function.py [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+    def copy(
+        self, new_name: Optional[str] = None, keep_variable_names: bool = False
+    ) -> "TrivialLoss":
+        return cast(
+            TrivialLoss,
+            super().copy(new_name=new_name, keep_variable_names=keep_variable_names),
+        )
+
+
+class RobustLoss(LossFunction):
```
Comment (mhmukadam) [inline theseus/core/loss_function.py:68]: Since you have abstract methods in this class
`class RobustLoss(LossFunction, abc.ABC)`

File: theseus/embodied/collision/collision.py [hunk 10]
```diff
@@ -23,11 +23,12 @@ def __init__(
         sdf_data: Variable,
         sdf_cell_size: Variable,
         cost_eps: Variable,
+        loss_function: Optional[LossFunction] = None,
```
Comment (mhmukadam) [inline theseus/embodied/collision/collision.py:26]: What's the behavior if we don't add this here. The super call would not have anything so the base class would assign trivial loss?

**Review by mhmukadam — COMMENTED** [review]
Looks good. We should add units tests before we merge to main.

File: theseus/core/loss_function.py (hunk 8 above)
Comment (fantaosha) [reply theseus/core/loss_function.py:31]: Will do.

File: theseus/embodied/collision/collision.py (hunk 10 above)
Comment (fantaosha) [reply theseus/embodied/collision/collision.py:26]: That's correct. We can skip `loss_function`, and then `collision` uses `TrivialLoss`.

**Comment by fantaosha** [comment]
> Looks good. We should add units tests before we merge to main.

@luisenp @mhmukadam Do you have any ideas about what unit tests we should have?

**Comment by mhmukadam** [comment]
> @luisenp @mhmukadam Do you have any ideas about what unit tests we should have?

You could add mock tests similar to ones for cost function, and also augment the current tests for objective, etc to account for the changes (checking add/erase, etc). Maybe Luis also has some suggestions.

**Comment by fantaosha** [comment]
Merged to `experimental`.
